Re: Homework 2 + Homework 4

batteryfree opened the pull request:
homework4
https://batteryfree.github.io/frontend-course-2019/homeworks/roman.vlasov_batteryfree/homework_4/dist/
homework2
https://batteryfree.github.io/frontend-course-2019/homeworks/roman.vlasov_batteryfree/homework_2/

batteryfree committed:
added homework-2

batteryfree committed:
Merge branch 'master' into hw-2

batteryfree committed:
Update app.js

Fixed regex for validating First Name and Last Name

batteryfree committed:
Merge branch 'master' into hw-2

batteryfree committed:
Merge branch 'master' into hw-2

batteryfree committed:
Merge branch 'master' into hw-2

smithua committed:
Merge branch 'master' into hw-2

batteryfree committed:
edit app.js

batteryfree committed:
Merge branch 'master' into hw-2

batteryfree committed:
Update app.js

added const elem

batteryfree committed:
Update style.css

batteryfree committed:
Update index.html

batteryfree committed:
Merge branch 'master' into hw-2

smithua committed:
Merge branch 'master' into hw-2

batteryfree committed:
Merge branch 'master' into hw-2

batteryfree committed:
Merge branch 'master' into gh-pages

batteryfree committed:
Merge branch 'master' into gh-pages

batteryfree committed:
Merge branch 'master' into gh-pages

TArch64 committed:
Merge branch 'master' into gh-pages

TArch64 commented on homeworks/roman.vlasov_batteryfree/homework_2/app.js:
```diff
@@ -0,0 +1,133 @@
+class SendMessage {
+    constructor(el) {
+        this.el = el;
+        this.validation = true;
```
validation це більше про процес. Тут підійде краще isValid, або деякі просто роблять validationError з стрінгою і якщо там щось є, то форма не валідна

batteryfree replied:
fixed

TArch64 commented on homeworks/roman.vlasov_batteryfree/homework_2/app.js:
```diff
@@ -0,0 +1,133 @@
+    }
+
+    listenEvents() {
+        this.elements.send.addEventListener('click', () => {
```
Якщо елемент використовується один раз (як у тебе на ініті), то його немає сенсу тримати в пам'яті. І краще слухай сабміт у форми, а не клік на кнопку. Тоді більше є варіантів як відправити запит (наприклад ентером)

batteryfree replied:
fixed

TArch64 commented on homeworks/roman.vlasov_batteryfree/homework_2/app.js:
```diff
@@ -0,0 +1,133 @@
+                    if (element.type.toLowerCase() === 'email') {
+                        elem.value = 'Email Address.....';
+                    }
+                }
```
Для всього що ти тут робиш придумати атрибут placeholdet в інпута

TArch64 commented on homeworks/roman.vlasov_batteryfree/homework_2/app.js:
```diff
@@ -0,0 +1,133 @@
+        this.elements = {
+            send: this.el.querySelector('.contact__btn'),
+            message: this.el.querySelector('.contact__txtarea'),
+            editsText: this.el.querySelectorAll('.contact__edit'),
```
Не роби так. Це не дуже зрочно. Зробюи на кожен інпут окремий клас і шукай кожен інпут окремо. І якщо ти щось дістаєш з дому по класу, то цей клас скоріш всього має бути унікальним  і до класів, які використовуються як селектори в джеес додають префікс js- щоб люди розуміли, що на ньому тримається логіка і ніхто нічого не замал при рефакторингу стилів

batteryfree replied:
fixed

TArch64 commented on homeworks/roman.vlasov_batteryfree/homework_2/app.js:
```diff
@@ -0,0 +1,133 @@
+        if (!this.validation) {
+            elem.style.border = '3px solid #F00';
+        } else {
+            elem.style.border = '1px solid #a9a9a9';
```
Замість того щоб змінювати інлайнові стиля навішук клас, а там хай уже css розбирайється сам як його показати. classList.toggle тобі допоможе

batteryfree replied:
fixed

TArch64 commented on homeworks/roman.vlasov_batteryfree/homework_2/app.js:
```diff
@@ -0,0 +1,133 @@
+document.addEventListener('DOMContentLoaded', () => {
+    const sendMessage = new SendMessage(document.querySelector('.contact__message'));
+    console.dir(sendMessage);
+});
```
Щоб не слухати івенти просто додай атрибут defer до script тегу, де в тебе цей файл підключається

batteryfree replied:
fixed

TArch64 commented on homeworks/roman.vlasov_batteryfree/homework_2/index.html:
```diff
@@ -0,0 +1,67 @@
+                <path d="M16 1l-16 16 3 3 3-3v13h8v-6h4v6h8v-13l3 3 3-3-16-16zM16 14c-1.105 0-2-0.895-2-2s0.895-2 2-2c1.105 0 2 0.895 2 2s-0.895 2-2 2z"></path>
+            </symbol>
+        </svg>
+        <script src="app.js"></script>
```
<script src="app.js" defer></script>

batteryfree replied:
fixed

TArch64 commented on homeworks/roman.vlasov_batteryfree/homework_2/app.js:
```diff
@@ -0,0 +1,133 @@
+            if (element.type.toLowerCase() === 'text') {
+                if (!VALID_NAME.test(element.value)) {
+                    this.validation = false;
+                    this.changeStyleInput(element);
```
Для того щоб показувати стан валідації філда в реальному часі люди використовують change або input івент в інпутів

batteryfree replied:
є питання з приводу івенту чендж в інпут майл ... в хромі він не відпрацйовуе якщо відразу надрукувати тільки прогонии та пертейти в інший філд ... чи можна зцим якось боротись ..чи доведеться мириться з цим?

TArch64 reviewed (changes requested):
То тобі на перший час. Як пофіксиш -- далі підемо

TArch64 committed:
Merge branch 'master' into gh-pages

TArch64 committed:
Merge branch 'master' into gh-pages

batteryfree committed:
Merge branch 'master' into gh-pages

batteryfree committed:
fixed app.js style.css index.html

batteryfree committed:
Merge branch 'master' into gh-pages

TArch64 commented on homeworks/roman.vlasov_batteryfree/homework_2/app.js:
```diff
@@ -0,0 +1,72 @@
+    listenEvents() {
+        const VALID_NAME = /^([a-zA-Z])+\s{1}([a-zA-Z])+$/;
+        const VALID_MAIL = /^([a-z0-9_-])+@[a-z0-9-]+\.([a-z]{2,4}\.)?[a-z]{2,4}$/i;
+        const VALID_MESS = /^/i;
```
Не всі валідації робляться регулярками. Тут регулярка прям зовсім оверхед `const VALID_MESS = /^/i;`

batteryfree replied:
fixed

TArch64 commented on homeworks/roman.vlasov_batteryfree/homework_2/app.js:
```diff
@@ -0,0 +1,72 @@
+            contactMessage: this.el.querySelector('.js-contact-message'),
+            contactName: this.el.querySelector('.js-contact-name'),
+            contactEmail: this.el.querySelector('.js-contact-email'),
+            contactSubmit: this.el.querySelector('.js-submit'),
```
слухай сабміт івент форми, а не кнопку. Для юзера так набагато зручніше. Люди які сабмітять форми ентером будуть кожен раз відчувати біль

batteryfree replied:
fixed

TArch64 commented on homeworks/roman.vlasov_batteryfree/homework_2/app.js:
```diff
@@ -0,0 +1,72 @@
+        }
+        this.setStyleInValid(element, false);
+        return true;
+    }
```
У тебе цей метод називається getValidateData, а насправді він робить валідацію, скидає навіщось element.value в пустий та рендерить втиху інвалідний стан стилів. Не роби так. Розбивай функції за тим що вони роблять. Чим менше задач виконує функція, тим краще.\

Універсальний валідатор ненапишеш, а в цьому ніхто розбиратись не захоче. Зроби окремо для кожного поля. Максимум що зазвичай роблять, так це виносять деякі елементи в хелпери. типу такого
```js
class SendMessage {
  isEmptyField(input) {
    return !input.value.trim()
  }

  isEmail(input) {
    return EMAIL_PATTERN.test(input.value);
  }

  validateEmail() {
    const emailEl = this.elements.contactEmail;

    if ( isEmptyField(emailEl) ) return `Email field is required';
    if ( isEmail(emailEl) ) return 'Email has not valid format';
  }

  validate() {
    const validationError = this.validateEmail() || this.validateName();
    const isValid = !validationError;
  }
}

```

batteryfree replied:
Вроде понял...спасибо!

TArch64 commented on homeworks/roman.vlasov_batteryfree/homework_2/app.js:
```diff
@@ -0,0 +1,72 @@
+}
+
+document.addEventListener('DOMContentLoaded', () => {
+    const sendMessage = new SendMessage(document.querySelector('.contact__message'));
```
ти тут забув про префікс для цього класу. І у тебе там скріпт через defer підключений. `document.addEventListener('DOMContentLoaded'` тут зайвий

batteryfree replied:
Да.. забыл).. спасибо

TArch64 committed:
Merge branch 'master' into gh-pages

TArch64 committed:
Merge branch 'master' into gh-pages

batteryfree committed:
added hw-4

batteryfree committed:
disable eslint in file main.js

batteryfree committed:
disabled lintric from first line validator.js

batteryfree committed:
fix style

batteryfree committed:
.added class js-contact__form && removed class js-submin in file index.html
.the contactSabmit variable/object has been renamed to contactForm, the event of the "contactForm" object has been changed click to submit

batteryfree committed:
fixed all

TArch64 commented:
Розбий на кожну домашку на окремий пулл реквест. Бо так цей буде перевірятись вічність

batteryfree commented:
Разбил..RE: RE HomeWork-2 #82...эт всё из-за пейджес...я уже поднимал вопрос поводу этого...и я сделал как понял...надо бы что-бы еще раз объяснили для тех кто в танке(то есть для меня )
